File: scripts/chapter_e2e_prompt_audit.ts
/**
 * Build final T2I / I2V prompts from chapter-analyze JSON (same helpers as LongVideoCreateView).
 * Usage: npx --yes tsx scripts/chapter_e2e_prompt_audit.ts <analyze.json>
 */
import { readFileSync } from 'node:fs';

import type { LongVideoCharacter, LongVideoScene, LongVideoShotState } from '../frontend/src/types';
import {
  keyframeGenerationPrompt,
  keyframePromptContextForShot,
  segmentVideoSubmitPreview,
  shotKeyframeText,
  shotNeedsKeyframe,
  shotVideoPrompt,
} from '../frontend/src/utils/longVideoProject';

type AnalyzeShot = {
  id?: string;
  order?: number;
  visual_prompt?: string;
  motion_prompt?: string;
  video_prompt?: string;
  start_visual_prompt?: string;
  anchor_visual_prompt?: string;
  segment_role?: LongVideoShotState['segment_role'];
  start_frame_mode?: LongVideoShotState['start_frame_mode'];
  segment_group_id?: string;
  segment_group_index?: number;
  face_anchor_shot_id?: string;
  flf_mode?: LongVideoShotState['flf_mode'];
  scene_prompt?: string;
  cast_looks?: LongVideoShotState['cast_looks'];
  scene_look?: LongVideoShotState['scene_look'];
  duration_sec?: number;
  first_frame_visibility?: LongVideoShotState['first_frame_visibility'];
  end_visibility?: LongVideoShotState['end_visibility'];
  characters_on_screen?: string[];
  first_frame_requirement?: string;
  location?: string;
  narrative_beat_index?: number;
  shot_size?: string;
  camera_zone_id?: string;
  first_frame_strategy?: LongVideoShotState['first_frame_strategy'];
};

type AnalyzePayload = {
  chapter_title?: string;
  synopsis?: string;
  mood?: string;
  character_anchor?: string;
  style_anchor?: string;
  characters?: LongVideoCharacter[];
  scenes?: LongVideoScene[];
  shots?: AnalyzeShot[];
  quality_issues?: { code: string; message: string; severity?: string }[];
};

function shotsFromApi(apiShots: AnalyzeShot[]): LongVideoShotState[] {
  return apiShots.map((s, i) => {
    const videoPrompt = (s.video_prompt || s.motion_prompt || '').trim();
    const startVisual = (s.start_visual_prompt || s.visual_prompt || '').trim();
    return {
      id: s.id || `shot_${String(i).padStart(2, '0')}`,
      order: i,
      visual_prompt: startVisual,
      motion_prompt: videoPrompt,
      video_prompt: videoPrompt,
      start_visual_prompt: startVisual || undefined,
      anchor_visual_prompt: s.anchor_visual_prompt?.trim() || undefined,
      segment_role: s.segment_role ?? 'keyframe',
      start_frame_mode: s.start_frame_mode ?? 'keyframe',
      segment_group_id: s.segment_group_id,
      segment_group_index: s.segment_group_index,
      face_anchor_shot_id: s.face_anchor_shot_id,
      flf_mode: s.flf_mode ?? 'none',
      scene_prompt: s.scene_prompt || '',
      cast_looks: s.cast_looks ?? [],
      scene_look: s.scene_look,
      duration_sec: s.duration_sec,
      first_frame_visibility: s.first_frame_visibility,
      end_visibility: s.end_visibility,
      characters_on_screen: s.characters_on_screen ?? [],
      first_frame_requirement: s.first_frame_requirement,
      location: s.location?.trim() || undefined,
      narrative_beat_index: s.narrative_beat_index,
      shot_size: s.shot_size?.trim() || undefined,
      camera_zone_id: s.camera_zone_id,
      first_frame_strategy: s.first_frame_strategy,
      status: 'draft' as const,
    };
  });
}

function main() {
  const path = process.argv[2];
  if (!path) {
    console.error('usage: tsx scripts/chapter_e2e_prompt_audit.ts <analyze.json>');
    process.exit(1);
  }
  const data = JSON.parse(readFileSync(path, 'utf8')) as AnalyzePayload;
  const shots = shotsFromApi(data.shots ?? []);
  const project = {
    character_anchor: data.character_anchor ?? '',
    style_anchor: data.style_anchor ?? '',
    characters: data.characters ?? [],
    scenes: data.scenes ?? [],
    shots,
    chain_mode: 'keyframe_only' as const,
  };

  console.log(`# ${data.chapter_title ?? 'chapter'}`);
  console.log(`shots=${shots.length} characters=${project.characters.length} scenes=${project.scenes.length}`);
  if (data.quality_issues?.length) {
    const codes = [...new Set(data.quality_issues.map((q) => q.code))];
    console.log(`quality_issue_codes=${codes.join(', ')}`);
  }
  console.log('');

  for (let i = 0; i < shots.length; i++) {
    const shot = shots[i];
    const label = `shot[${i}] #${i + 1} role=${shot.segment_role} group=${shot.segment_group_id ?? '-'}`;
    console.log(`## ${label}`);
    console.log(`location: ${shot.location ?? '(none)'}`);
    console.log(`scene_prompt: ${shot.scene_prompt ?? ''}`);
    console.log(`first_frame_requirement: ${shot.first_frame_requirement ?? ''}`);
    console.log(`characters_on_screen: ${(shot.characters_on_screen ?? []).join(', ') || '(none)'}`);
    if (shot.segment_role === 'face_anchor') {
      console.log(`anchor_visual_prompt: ${shot.anchor_visual_prompt ?? ''}`);
    } else {
      console.log(`start_visual_prompt: ${shot.start_visual_prompt ?? shot.visual_prompt ?? ''}`);
    }
    console.log(`video_prompt: ${shotVideoPrompt(shot)}`);

    if (shotNeedsKeyframe(shot)) {
      const visual = shotKeyframeText(shot);
      const t2i = keyframeGenerationPrompt(visual, keyframePromptContextForShot(shot, project));
      console.log('');
      console.log('### T2I submit prompt');
      console.log(t2i);
    } else {
      console.log('');
      console.log('### T2I: skipped (no keyframe needed)');
    }

    const i2vPreview = segmentVideoSubmitPreview(shot, shots, { shotIndex: i, chainMode: 'keyframe_only' });
    console.log('');
    console.log('### I2V submit preview');
    console.log(i2vPreview || '(empty — needs keyframe_asset_id at generation time)');
    console.log('');
  }
}

main();

File: scripts/chapter_e2e_prompt_audit.py
#!/usr/bin/env python3
"""Run chapter parse on benchmark fixtures and audit T2I/I2V prompts."""
from __future__ import annotations

import argparse
import json
import subprocess
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from backend.core.contracts import ScriptParseDecomposeRequest, ScriptParseExpandRequest
from tests.chapter_parse_benchmark_cases import CASE_BY_ID, CHAPTER_PARSE_BENCHMARK_CASES
from tests.script_parse_integration import _load_llm_service

OUT_DIR = ROOT / "tests" / "benchmark" / "outputs" / "chapter_e2e"


def _response_to_dict(resp) -> dict:
    if hasattr(resp, "model_dump"):
        return resp.model_dump()
    return dict(resp)


def _check_shot_fields(shot: dict, index: int) -> list[str]:
    """Generic policy checks (no story-specific strings)."""
    issues: list[str] = []
    from backend.engine.llm.storyboard_cast import find_name_look_tags

    for fld in ("scene_prompt", "start_visual_prompt", "visual_prompt", "anchor_visual_prompt", "video_prompt"):
        text = str(shot.get(fld) or "")
        if find_name_look_tags(text):
            issues.append(f"shot {index}: {fld} has inline Name（…） look tag")
    role = shot.get("segment_role")
    if role == "face_anchor":
        av = str(shot.get("anchor_visual_prompt") or shot.get("start_visual_prompt") or "").strip()
        if not av:
            issues.append(f"shot {index}: face_anchor missing anchor/start visual")
        loc = str(shot.get("location") or "").strip()
        if loc and av and loc not in av and loc not in str(shot.get("scene_prompt") or ""):
            issues.append(f"shot {index}: face_anchor visual may omit location {loc!r}")
    vp = str(shot.get("video_prompt") or shot.get("motion_prompt") or "").strip()
    if not vp:
        issues.append(f"shot {index}: empty video_prompt")
    return issues


def run_case(case_id: str, *, skip_tsx: bool = False) -> int:
    case = CASE_BY_ID[case_id]
    svc = _load_llm_service()
    if not svc.is_available():
        print("SKIP: local LLM not available", file=sys.stderr)
        return 2

    print(f"\n{'=' * 72}\nCASE {case_id} — {case.title}\n{'=' * 72}")
    t0 = time.perf_counter()
    decomposed = svc.script_parse_decompose(
        ScriptParseDecomposeRequest(
            script_text=case.load_script(),
            title=case.title,
            locale=case.locale,
        )
    )
    resp = svc.script_parse_expand(
        ScriptParseExpandRequest(
            script_artifact=decomposed.script_artifact,
            locale=case.locale,
            target_duration_sec=case.target_duration_sec,
            segment_duration_sec=case.segment_duration_sec,
        )
    )
    elapsed = time.perf_counter() - t0
    payload = _response_to_dict(resp)
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    out_json = OUT_DIR / f"{case_id}_analyze.json"
    out_json.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")

    shots = payload.get("shots") or []
    beats = payload.get("scene_beats") or []
    quality = payload.get("quality_issues") or []
    print(f"elapsed={elapsed:.1f}s beats={len(beats)} shots={len(shots)} llm_calls={payload.get('llm_calls')}")
    print(f"saved: {out_json.relative_to(ROOT)}")

    field_issues: list[str] = []
    for i, shot in enumerate(shots):
        field_issues.extend(_check_shot_fields(shot, i))
    if field_issues:
        print("\nFIELD CHECK WARNINGS:")
        for msg in field_issues[:20]:
            print(f"  - {msg}")
        if len(field_issues) > 20:
            print(f"  ... and {len(field_issues) - 20} more")
    else:
        print("\nFIELD CHECK: ok (no inline look tags, video_prompt present)")

    if quality:
        codes = sorted({q.get("code") for q in quality if q.get("code")})
        print(f"quality_issues ({len(quality)}): {', '.join(codes)}")

    if skip_tsx:
        return 1 if field_issues else 0

    print("\n--- Final T2I / I2V prompts (frontend helpers) ---\n")
    proc = subprocess.run(
        ["npx", "--yes", "tsx", str(ROOT / "scripts" / "chapter_e2e_prompt_audit.ts"), str(out_json)],
        cwd=ROOT / "frontend",
        capture_output=True,
        text=True,
    )
    if proc.stdout:
        print(proc.stdout)
    if proc.returncode != 0:
        print(proc.stderr, file=sys.stderr)
        return proc.returncode
    return 1 if field_issues else 0


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--case", choices=("wukong", "rainy_night", "all"), default="all")
    parser.add_argument("--skip-tsx", action="store_true")
    args = parser.parse_args()
    case_ids = [c.case_id for c in CHAPTER_PARSE_BENCHMARK_CASES] if args.case == "all" else [args.case]
    rc = 0
    for cid in case_ids:
        code = run_case(cid, skip_tsx=args.skip_tsx)
        rc = max(rc, code)
    return rc


if __name__ == "__main__":
    raise SystemExit(main())
